tsl/src/compression/compression_scankey.c

```diff
@@ -322,10 +322,12 @@ build_index_scankeys_using_slot(Oid hypertable_relid, Relation in_rel, Relation
 			const NameData *attname = attnumAttName(in_rel, in_attnum);
 			AttrNumber column_attno = get_attnum(out_rel->rd_id, NameStr(*attname));
 
-			/* Make sure we find columns in key columns in order to select the right index */
+			/* Make sure we find columns in key columns in order to select the right index
+			 * We skip over any non-key columns
+			 */
 			if (!bms_is_member(column_attno, key_columns))
 			{
-				break;
+				continue;
```

Comment:
Will this really work? Don't you need at least the first index column as a scan key?

Reply:
There is a check later for number of scankeys a particular index can use, we only use the index if we can generate at least one scan key.

Reply:
But we can't use an index if we only match, e.g., the last index key right?

With this check you might generate a suffix.

Let's say an index includes a,b,c

then you can use it if you match <a>, <a,b>, <a,b,c>, or <a,c> right?

But you cannot use the index if you match <b> or <c> or <b,c>.... or am I missing something?

Here it looks like you can generate a match for the latter examples too since you can skip any column, including the first one.

Reply:
Ideally, we should also look for longest prefix match. Here it looks like we are happy with first matching index even if there is a better one that matches more keys.

Maybe the simple approach works for our case because we know what indexes exist, but I am not sure the code is really doing the right thing here.

Reply:
Even though not optimal, there are cirumstances where picking an index without the leading edge is actually done in Postgres?

```
feike=# \d a
                 Table "public.a"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 odd    | boolean |           |          |
 j      | integer |           |          |
 t      | text    |           |          |
Indexes:
    "a_odd_j_idx" btree (odd, j)

feike=# explain select * from a where j=4;
                                 QUERY PLAN
----------------------------------------------------------------------------
 Index Scan using a_odd_j_idx on a  (cost=0.42..18484.43 rows=1 width=1909)
   Index Cond: (j = 4)
(2 rows)
```

It's more a statistics based algorithm that should pick the index though?

Reply:
@feikesteenbergen but isn't this a contrived example by using a boolean (two possible values) as the first index key? I doubt that this is a common case in this situation.

In any case, we should look for the longest-prefix matching index and not just pick the first index that matches at least on index key.

Reply:
>It's more a statistics based algorithm that should pick the index though?

Also, this index matching code isn't using stats at all.

Reply:
FWIW,  If you are confident that the current code is OK despite the concerns/questions I raised, then feel free to go ahead.

Reply:
> But we can't use an index if we only match, e.g., the last index key right?

It's going to return the correct results. It'll be a kind of "index seq scan", where you scan an index but don't actually have any search keys, only plain scan keys ("index cond" vs "index filter" in explain). We actually had a bug like this in our catalog lookups where we used a wrong index.